### PR TITLE
Add user <-> public repo mapping table

### PR DESCRIPTION
**Pull request by arussellsaw**
This PR adds a new table to map users to public repositories, this is so we can allow users to add public repositories on sourcegraph.com without configuring a code host, and have that repo appear within their default search context.

### CODE REVIEW

File: internal/db/user_public_repos.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+// initialized it.
+func (s *UserPublicRepoStore) ensureStore() {
+	s.once.Do(func() {
+		if s.store == nil {
+			s.store = basestore.NewWithDB(dbconn.Global, sql.TxOptions{})
+		}
+	})
+}
+
+func (s *UserPublicRepoStore) SetUserRepo(ctx context.Context, userID int32, repoID api.RepoID) error {
```
Comment (tsenart) [inline internal/db/user_public_repos.go:38]: How is this going to be used? This API makes it seem like we can only associate already synced repos, but the UI asks the user for repo names. Can you describe the steps?

**Review by tsenart — APPROVED** [review]

File: internal/db/user_public_repos.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	s.ensureStore()
+	_, err := s.store.Query(ctx, sqlf.Sprintf("INSERT INTO user_public_repos(user_id, repo_id) VALUES %v %v", userID, repoID))
```
Comment (ryanslade) [inline internal/db/user_public_repos.go:40]: You could use `s.store.Exec` here since you don't expect to return any rows

**Review by ryanslade — APPROVED** [review]

**Comment by ryanslade** [comment]
Oh, you should probably add a test

File: internal/db/user_public_repos.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+	s.once.Do(func() {
+		if s.store == nil {
+			s.store = basestore.NewWithDB(dbconn.Global, sql.TxOptions{})
+		}
+	})
+}
+
+func (s *UserPublicRepoStore) SetUserRepo(ctx context.Context, userID int32, repoID api.RepoID) error {
+	s.ensureStore()
+	return s.store.Exec(ctx, sqlf.Sprintf("INSERT INTO user_public_repos(user_id, repo_id) VALUES %v %v", userID, repoID))
```
Comment (ryanslade) [inline internal/db/user_public_repos.go:40]:
```suggestion
	return s.store.Exec(ctx, sqlf.Sprintf("INSERT INTO user_public_repos(user_id, repo_id) VALUES (%v, %v)", userID, repoID))
```

**Review by ryanslade — CHANGES_REQUESTED** [review]
Noticed an error in the query and the need for some tests

**Comment by arussellsaw** [comment]
@ryanslade good call, done. i also added a read method so i could actually test it

File: internal/db/user_public_repos.go [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+func (s *UserPublicRepoStore) ListByUser(ctx context.Context, userID int32) ([]int32, error) {
+	s.ensureStore()
+
+	rows, err := s.store.Query(ctx, sqlf.Sprintf("SELECT repo_id FROM user_public_repos WHERE user_id = %v", userID))
+	if err != nil {
+		return nil, err
+	}
+	out := []int32{}
```
Comment (ryanslade) [inline internal/db/user_public_repos.go:51]: We have `basestore.ScanInts` already, maybe make a `ScanInt32s`?

Comment (ryanslade) [reply internal/db/user_public_repos.go:51]: Then you can just write:

```go
out, err := basestore.ScanInt32s(s.store.Query(...))
```

File: internal/db/user_public_repos_test.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
+package db
+
+import (
+	"context"
+	"github.com/sourcegraph/sourcegraph/internal/db/dbtesting"
```
Comment (ryanslade) [inline internal/db/user_public_repos_test.go:5]: Please group the imports

**Review by ryanslade — APPROVED** [review]

File: internal/db/user_public_repos.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+		if s.store == nil {
+			s.store = basestore.NewWithDB(dbconn.Global, sql.TxOptions{})
+		}
+	})
+}
+
+func (s *UserPublicRepoStore) SetUserRepo(ctx context.Context, userID int32, repoID api.RepoID) error {
+	s.ensureStore()
+
+	return s.store.Exec(ctx, sqlf.Sprintf("INSERT INTO user_public_repos(user_id, repo_id) VALUES (%v, %v)", userID, repoID))
```
Comment (unknwon) [inline internal/db/user_public_repos.go:41]: We don't have a unique constraint on `(user_id, repo_id)` so it's possible that we have duplicated pairs in this table?

If we have unique constraint on `(user_id, repo_id)`, then we can easily add `ON CONFLICT DO NOTHING` in the above SQL. WDYT?

Comment (ryanslade) [reply internal/db/user_public_repos.go:41]: Good point! Please add that @arussellsaw 

This is something we forgot to do for the `external_service_repos` table originally and it was a pain to add later.

**Review by ryanslade — APPROVED** [review]
LGTM

Would still like the ScanInt32s addition but that can also be done in another PR

**Comment by arussellsaw** [comment]
i was feeling lazy yesterday, i've added the scan func now

File: internal/db/user_public_repos.go [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+func (s *UserPublicRepoStore) ListByUser(ctx context.Context, userID int32) ([]int32, error) {
+	s.ensureStore()
+
+	rows, err := basestore.ScanInt32s(s.store.Query(ctx, sqlf.Sprintf("SELECT repo_id FROM user_public_repos WHERE user_id = %v", userID)))
```
Comment (ryanslade) [inline internal/db/user_public_repos.go:47]: nit: You could just return here:

**Pull request merged**